### ISSUE TITLE
clusterctl delete cluster appears to hang on 0.1.4

### ISSUE THREAD

**Issue by akutz** [kubernetes-sigs/cluster-api#1105]
/kind bug

**What steps did you take and what happened:**
`clusterctl delete cluster` hangs indefinitely waiting for a machine to be deleted

```shell
0702 16:23:02.071499   21340 clusterclient.go:870] Waiting for Machines to be deleted...
I0702 16:23:02.074064   21340 round_trippers.go:438] GET https://localhost:41079/apis/cluster.k8s.io/v1alpha1/namespaces/default/machines 200 OK in 2 milliseconds
I0702 16:23:12.074585   21340 clusterclient.go:870] Waiting for Machines to be deleted...
I0702 16:23:12.078698   21340 round_trippers.go:438] GET https://localhost:41079/apis/cluster.k8s.io/v1alpha1/namespaces/default/machines 200 OK in 3 milliseconds
I0702 16:23:22.074569   21340 clusterclient.go:870] Waiting for Machines to be deleted...
I0702 16:23:22.077669   21340 round_trippers.go:438] GET https://localhost:41079/apis/cluster.k8s.io/v1alpha1/namespaces/default/machines 200 OK in 2 milliseconds
I0702 16:23:32.074524   21340 clusterclient.go:870] Waiting for Machines to be deleted...
I0702 16:23:32.077266   21340 round_trippers.go:438] GET https://localhost:41079/apis/cluster.k8s.io/v1alpha1/namespaces/default/machines 200 OK in 2 milliseconds
I0702 16:23:42.074591   21340 clusterclient.go:870] Waiting for Machines to be deleted...
I0702 16:23:42.077924   21340 round_trippers.go:438] GET https://localhost:41079/apis/cluster.k8s.io/v1alpha1/namespaces/default/machines 200 OK in 3 milliseconds
I0702 16:23:52.074538   21340 clusterclient.go:870] Waiting for Machines to be deleted...
I0702 16:23:52.077611   21340 round_trippers.go:438] GET https://localhost:41079/apis/cluster.k8s.io/v1alpha1/namespaces/default/machines 200 OK in 2 milliseconds
I0702 16:24:02.074546   21340 clusterclient.go:870] Waiting for Machines to be deleted...
I0702 16:24:02.078224   21340 round_trippers.go:438] GET https://localhost:41079/apis/cluster.k8s.io/v1alpha1/namespaces/default/machines 200 OK in 3 milliseconds
I0702 16:24:12.074576   21340 clusterclient.go:870] Waiting for Machines to be deleted...
I0702 16:24:12.077736   21340 round_trippers.go:438] GET https://localhost:41079/apis/cluster.k8s.io/v1alpha1/namespaces/default/machines 200 OK in 3 milliseconds
I0702 16:24:22.074522   21340 clusterclient.go:870] Waiting for Machines to be deleted...
I0702 16:24:22.077215   21340 round_trippers.go:438] GET https://localhost:41079/apis/cluster.k8s.io/v1alpha1/namespaces/default/machines 200 OK in 2 milliseconds
I0702 16:24:32.074529   21340 clusterclient.go:870] Waiting for Machines to be deleted...
I0702 16:24:32.077148   21340 round_trippers.go:438] GET https://localhost:41079/apis/cluster.k8s.io/v1alpha1/namespaces/default/machines 200 OK in 2 milliseconds
I0702 16:24:42.074562   21340 clusterclient.go:870] Waiting for Machines to be deleted...
I0702 16:24:42.077514   21340 round_trippers.go:438] GET https://localhost:41079/apis/cluster.k8s.io/v1alpha1/namespaces/default/machines 200 OK in 2 milliseconds
I0702 16:24:52.074539   21340 clusterclient.go:870] Waiting for Machines to be deleted...
I0702 16:24:52.077985   21340 round_trippers.go:4
```

**What did you expect to happen:**
For `clusterctl delete cluster` to not hang.

**Anything else you would like to add:**
The CAPV manager log has the following, ad infinitum:

```shell
E0702 21:24:09.444342       1 machine_controller.go:186] Error deleting node "kutz-controlplane-1" for machine "kutz-controlplane-1": Delete https://192.168.5.175:6443/api/v1/nodes/kutz-controlplane-1: dial tcp 192.168.5.175:6443: i/o timeout
```

Based on the log message, the error appears to be occurring [here](https://github.com/kubernetes-sigs/cluster-api/blob/bf950adfea7c117e58ca30770d02067d8e43d3c7/pkg/controller/machine/machine_controller.go#L183-L189):

```golang
		klog.Infof("Reconciling machine %q triggers delete", name)
		if err := r.actuator.Delete(ctx, cluster, m); err != nil {
			if requeueErr, ok := errors.Cause(err).(controllerError.HasRequeueAfterError); ok {
				klog.Infof("Actuator returned requeue-after error: %v", requeueErr)
				return reconcile.Result{Requeue: true, RequeueAfter: requeueErr.GetRequeueAfter()}, nil
			}


			klog.Errorf("Failed to delete machine %q: %v", name, err)
			return reconcile.Result{}, err
		}


		if m.Status.NodeRef != nil {
			klog.Infof("Deleting node %q for machine %q", m.Status.NodeRef.Name, m.Name)
			if err := r.deleteNode(ctx, cluster, m.Status.NodeRef.Name); err != nil && !apierrors.IsNotFound(err) {
				klog.Errorf("Error deleting node %q for machine %q: %v", m.Status.NodeRef.Name, name, err)
				return reconcile.Result{}, err
			}
		}
```

Since this is an attempt to delete a single-node cluster that is the initial target cluster deployed by `clusterctl create cluster`, I think what is happening is an ordering issue.

The nodeRef will never be successfully deleted since the target cluster is already offline since the machine actuator's `delete` call removed the backend VM.

Thus all we see in the manager log is one timeout after another:

```shell
I0702 21:23:39.422211       1 machine_controller.go:115] Reconciling Machine "kutz-controlplane-1"
I0702 21:23:39.422268       1 machine_controller.go:172] Reconciling machine "kutz-controlplane-1" triggers delete
I0702 21:23:39.427235       1 actuator.go:140] default-logger/cluster.k8s.io/v1alpha1/default/kutz/kutz-controlplane-1 "level"=2 "msg"="deleting machine"  
I0702 21:23:39.429579       1 util.go:183] default-logger/cluster.k8s.io/v1alpha1/default/kutz/kutz-controlplane-1/task-216212 "level"=4 "msg"="task found"  "description-id"="VirtualMachine.destroy" "state"="success"
I0702 21:23:39.429609       1 util.go:192] default-logger/cluster.k8s.io/v1alpha1/default/kutz/kutz-controlplane-1/task-216212 "level"=4 "msg"="task is a success"  "description-id"="VirtualMachine.destroy"
I0702 21:23:39.432616       1 machine_context.go:270] default-logger/cluster.k8s.io/v1alpha1/default/kutz/kutz-controlplane-1/task-216212 "level"=6 "msg"="generated json patch for object"  "json-patch"="[\n  {\n    \"op\": \"remove\",\n    \"path\": \"/spec/providerSpec/value/machineRef\"\n  }\n]"
I0702 21:23:39.438498       1 round_trippers.go:438] PATCH https://10.96.0.1:443/apis/cluster.k8s.io/v1alpha1/namespaces/default/machines/kutz-controlplane-1 200 OK in 5 milliseconds
I0702 21:23:39.438676       1 machine_context.go:277] default-logger/cluster.k8s.io/v1alpha1/default/kutz/kutz-controlplane-1/task-216212 "level"=6 "msg"="patch object success"  
I0702 21:23:39.442169       1 round_trippers.go:438] POST https://10.96.0.1:443/api/v1/namespaces/default/events 201 Created in 3 milliseconds
I0702 21:23:39.442733       1 round_trippers.go:438] PUT https://10.96.0.1:443/apis/cluster.k8s.io/v1alpha1/namespaces/default/machines/kutz-controlplane-1/status 200 OK in 3 milliseconds
I0702 21:23:39.442874       1 machine_context.go:300] default-logger/cluster.k8s.io/v1alpha1/default/kutz/kutz-controlplane-1/task-216212 "level"=6 "msg"="patch status success"  
I0702 21:23:39.442914       1 machine_controller.go:184] Deleting node "kutz-controlplane-1" for machine "kutz-controlplane-1"
I0702 21:23:39.445721       1 round_trippers.go:438] PATCH https://10.96.0.1:443/api/v1/namespaces/default/events/kutz-controlplane-1.15adb404d3cde107 200 OK in 2 milliseconds
I0702 21:23:39.447453       1 round_trippers.go:438] POST https://10.96.0.1:443/api/v1/namespaces/default/events 201 Created in 1 milliseconds
I0702 21:23:43.506845       1 round_trippers.go:438] GET https://192.168.5.175:6443/api/v1/nodes  in 30000 milliseconds
E0702 21:23:43.507034       1 actuator.go:152] [cluster-actuator]/cluster.k8s.io/v1alpha1/default/kutz "msg"="unable to list nodes for target cluster" "error"="Get https://192.168.5.175:6443/api/v1/nodes: dial tcp 192.168.5.175:6443: i/o timeout"  
I0702 21:23:43.515226       1 round_trippers.go:438] PUT https://10.96.0.1:443/apis/cluster.k8s.io/v1alpha1/namespaces/default/clusters/kutz/status 200 OK in 5 milliseconds
I0702 21:23:43.516064       1 cluster_context.go:361] [cluster-actuator]/cluster.k8s.io/v1alpha1/default/kutz "level"=6 "msg"="patch status success"  
I0702 21:23:43.516164       1 cluster_controller.go:143] Actuator returned requeue-after error: requeue in: 20s
I0702 21:23:43.519861       1 round_trippers.go:438] PATCH https://10.96.0.1:443/api/v1/namespaces/default/events/kutz.15adb403a46284ca 200 OK in 3 milliseconds
I0702 21:23:43.523229       1 round_trippers.go:438] PATCH https://10.96.0.1:443/api/v1/namespaces/default/events/kutz.15adb403a462b481 200 OK in 2 milliseconds
I0702 21:24:03.516480       1 cluster_controller.go:91] Running reconcile Cluster for "kutz"
I0702 21:24:03.516536       1 cluster_controller.go:140] reconciling cluster object kutz triggers idempotent reconcile.
I0702 21:24:03.518804       1 actuator.go:82] [cluster-actuator]/cluster.k8s.io/v1alpha1/default/kutz "level"=6 "msg"="reconciling cluster"  
I0702 21:24:03.518862       1 certificates.go:73] [cluster-actuator]/cluster.k8s.io/v1alpha1/default/kutz "level"=6 "msg"="reconciling certificates"  
I0702 21:24:09.444164       1 round_trippers.go:438] DELETE https://192.168.5.175:6443/api/v1/nodes/kutz-controlplane-1  in 30000 milliseconds
E0702 21:24:09.444342       1 machine_controller.go:186] Error deleting node "kutz-controlplane-1" for machine "kutz-controlplane-1": Delete https://192.168.5.175:6443/api/v1/nodes/kutz-controlplane-1: dial tcp 192.168.5.175:6443: i/o timeout
I0702 21:24:10.444776       1 machine_controller.go:115] Reconciling Machine "kutz-controlplane-1"
I0702 21:24:10.444883       1 machine_controller.go:172] Reconciling machine "kutz-controlplane-1" triggers delete
I0702 21:24:10.452061       1 actuator.go:140] default-logger/cluster.k8s.io/v1alpha1/default/kutz/kutz-controlplane-1 "level"=2 "msg"="deleting machine"  
I0702 21:24:10.454649       1 util.go:55] default-logger/cluster.k8s.io/v1alpha1/default/kutz/kutz-controlplane-1 "level"=6 "msg"="finding vm by instance UUID"  "instance-uuid"="7dd88196-9d0f-11e9-a430-0242ac110002"
I0702 21:24:10.458275       1 machine_controller.go:184] Deleting node "kutz-controlplane-1" for machine "kutz-controlplane-1"
I0702 21:24:10.462010       1 round_trippers.go:438] PATCH https://10.96.0.1:443/api/v1/namespaces/default/events/kutz-controlplane-1.15adb40e5699f56c 200 OK in 3 milliseconds
I0702 21:24:33.520791       1 round_trippers.go:438] GET https://192.168.5.175:6443/api/v1/nodes  in 30000 milliseconds
E0702 21:24:33.520896       1 actuator.go:152] [cluster-actuator]/cluster.k8s.io/v1alpha1/default/kutz "msg"="unable to list nodes for target cluster" "error"="Get https://192.168.5.175:6443/api/v1/nodes: dial tcp 192.168.5.175:6443: i/o timeout"  
I0702 21:24:33.527657       1 round_trippers.go:438] PUT https://10.96.0.1:443/apis/cluster.k8s.io/v1alpha1/namespaces/default/clusters/kutz/status 200 OK in 5 milliseconds
I0702 21:24:33.528414       1 cluster_context.go:361] [cluster-actuator]/cluster.k8s.io/v1alpha1/default/kutz "level"=6 "msg"="patch status success"  
I0702 21:24:33.528460       1 cluster_controller.go:143] Actuator returned requeue-after error: requeue in: 20s
I0702 21:24:33.532025       1 round_trippers.go:438] PATCH https://10.96.0.1:443/api/v1/namespaces/default/events/kutz.15adb403a46284ca 200 OK in 3 milliseconds
I0702 21:24:33.535348       1 round_trippers.go:438] PATCH https://10.96.0.1:443/api/v1/namespaces/default/events/kutz.15adb403a462b481 200 OK in 3 milliseconds
I0702 21:24:40.460207       1 round_trippers.go:438] DELETE https://192.168.5.175:6443/api/v1/nodes/kutz-controlplane-1  in 30000 milliseconds
E0702 21:24:40.460303       1 machine_controller.go:186] Error deleting node "kutz-controlplane-1" for machine "kutz-controlplane-1": Delete https://192.168.5.175:6443/api/v1/nodes/kutz-controlplane-1: dial tcp 192.168.5.175:6443: i/o timeout
```

**Environment:**
CAPV `master`

- Cluster-api version:
0.1.4

- Minikube/KIND version:
Kind 0.3.0

- Kubernetes version: (use `kubectl version`):
1.13.6

- OS (e.g. from `/etc/os-release`):
Ubuntu 18.04.2 LTS

cc @ncdc (since you appear to be having the same issue according to [this Slack comment](https://kubernetes.slack.com/archives/C8TSNPY4T/p1562103056148100?thread_ts=1562102887.147000&cid=C8TSNPY4T)

**Comment by akutz** [comment]
Hi @timothysc,

This will also need to be back ported to v1a1.

**Comment by ncdc** [comment]
Fixed by #1110 
/close
